File: CRISIS/CRISIS/Services/DashboardRepository.swift
import Foundation

struct DashboardRepository {
    private let metalService: MetalPriceService
    private let macroService: MacroIndicatorService
    private let crisisService: CrisisMonitorService
    private let staticSnapshot: DashboardSnapshot?

    init(
        metalService: MetalPriceService = MetalPriceService(),
        macroService: MacroIndicatorService = MacroIndicatorService(),
        crisisService: CrisisMonitorService = CrisisMonitorService(),
        mockData: DashboardSnapshot? = nil
    ) {
        self.metalService = metalService
        self.macroService = macroService
        self.crisisService = crisisService
        self.staticSnapshot = mockData
    }

    func makeSnapshot() async throws -> DashboardSnapshot {
        if let staticSnapshot {
            return staticSnapshot
        }

        async let metalsTask = metalService.fetchAssets()
        async let inflationTask = macroService.fetchIndicator(.inflation)
        async let growthTask = macroService.fetchIndicator(.growth)
        async let defenseTask = macroService.fetchIndicator(.defense)
        async let crisisTask = crisisService.fetchEvents()

        let metals = try await metalsTask
        let (inflation, inflationSeries) = try await inflationTask
        let (growth, growthSeries) = try await growthTask
        let (defense, defenseSeries) = try await defenseTask
        let crises = try await crisisTask

        let macroOverview = MacroOverview(indicators: [inflation, growth, defense])
        let macroSeries = [inflationSeries, growthSeries, defenseSeries]

        return DashboardSnapshot(
            metals: metals,
            macroOverview: macroOverview,
            macroSeries: macroSeries,
            crises: crises
        )
    }
}

File: android/core-data/src/main/java/de/vibecode/crisis/core/data/DashboardRepository.kt
package de.vibecode.crisis.core.data

import de.vibecode.crisis.core.model.DashboardSnapshot
import de.vibecode.crisis.core.model.MacroOverview
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

class DashboardRepository(
    private val metalService: MetalPriceService,
    private val macroService: MacroIndicatorService,
    private val crisisService: CrisisMonitorService,
    private val preferences: UserPreferences
) {
    suspend fun makeSnapshot(): DashboardSnapshot = coroutineScope {
        val metalsTask = async { metalService.fetchAssets() }
        val inflationTask = async { macroService.fetchIndicator(de.vibecode.crisis.core.model.MacroIndicatorKind.INFLATION) }
        val growthTask = async { macroService.fetchIndicator(de.vibecode.crisis.core.model.MacroIndicatorKind.GROWTH) }
        val defenseTask = async { macroService.fetchIndicator(de.vibecode.crisis.core.model.MacroIndicatorKind.DEFENSE) }
        val crisisTask = async {
            val settings = preferences.getCrisisSettings()
            crisisService.fetchEvents(settings)
        }

        val metals = metalsTask.await()
        val (inflation, inflationSeries) = inflationTask.await()
        val (growth, growthSeries) = growthTask.await()
        val (defense, defenseSeries) = defenseTask.await()
        val crises = crisisTask.await()

        DashboardSnapshot(
            metals = metals,
            macroOverview = MacroOverview(listOf(inflation, growth, defense)),
            macroSeries = listOf(inflationSeries, growthSeries, defenseSeries),
            crises = crises
        )
    }
}
